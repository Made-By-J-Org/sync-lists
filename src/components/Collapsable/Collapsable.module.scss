// I'd love to do it with before and after and proper unicode icons
// but text next to icon prevents me from doing it...
//summary::after {
//  content: "\02C4 Show";
//}
//summary::after {
//  content: "\02C5 Hide";
//}

@import '../../styles/colors';

details {
  cursor: pointer;

  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  &:not([open]) {
    animation-duration: .2s;
    animation-name: fold-in;
  }

  summary {
    display: flex;
    align-items: flex-start;
    justify-content: space-between;
    padding: 15px;
  }

  // Disable browser default marker
  summary::-webkit-details-marker,
  summary::marker {
    display: none;
    content: "";
  }
}

// Collapsible OPEN
details[open] {
  .content {
    animation-duration: .2s;
    animation-name: fold-out;
  }
}

// set class when open
.rotateIcon {
  > svg {
    transform: rotate(180deg);
  }
}

.positionText {
  width: 55px; // to stop text and icon from moving from right side.
  font-weight: lighter;
}

// align icon...
span:first-child {
  > svg {
    position: relative;
    top: 2px;
  }
}

.content {
  margin-top: 10px;
  margin-left: 25px;
}

@keyframes fold-out {
  0% {
    margin-top: -1rem;
    opacity: 0;
  }

  100% {
    margin-top: 15px;
    opacity: 1;
  }
}

@keyframes fold-in {
  0% {
    padding-bottom: 1rem;
  }

  100% {
    padding-bottom: 0;
  }
}
