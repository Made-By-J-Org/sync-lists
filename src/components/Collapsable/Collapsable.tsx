import React from 'react';

import {IconChevronDown, IconList} from '../Icons';
import TasksGroup from "../TasksGroup/TasksGroup";
import {ApiSchema} from "../../api/api.mock";

import styles from './Collapsable.module.scss';

interface CollapsableProps {
  tasksList: ApiSchema;
}

const Collapsable = ({tasksList}: CollapsableProps): React.ReactElement => {
  // due to 'show/hide' text I need a state.
  const [open, setOpen] = React.useState(false);

  return (
    <details
      onToggle={() => setOpen(prev => !prev)}
      data-testid='collapsible-details'
    >
      <summary>
        <span>
          <IconList/>
          {' '}
          <b>{tasksList.name}</b>
        </span>
        <span
          className={`
            ${open ? styles.rotateIcon : ''}
            ${styles.positionText}
          `}
        >
          <IconChevronDown/>
          {' '}
          {open ? 'Hide' : 'Show'}
        </span>
      </summary>

      <div className={styles.content}>
        <TasksGroup tasks={tasksList.tasks}/>
      </div>
    </details>
  );
};

export default Collapsable;
