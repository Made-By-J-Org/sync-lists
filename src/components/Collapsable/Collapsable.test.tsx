import React from 'react';
import {render, screen} from '@testing-library/react';
import Collapsable from './Collapsable';
import userEvent from '@testing-library/user-event'

describe('Collapsable', () => {
  const tasksList = {
    name: 'Test Tasks',
    tasks: [
      {
        checked: true,
        description: "Test Task",
        id: 1,
        value: 1
      },
      {
        checked: false,
        description: "Test Task2",
        id: 2,
        value: 2
      }
    ]
  };

  it('render component', () => {
    const {baseElement} = render(<Collapsable tasksList={tasksList}/>);
    expect(baseElement).toBeInTheDocument();
  });

  it('match snapshot', () => {
    const {asFragment} = render(<Collapsable tasksList={tasksList}/>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('toggles the state when details is clicked twice', async () => {
    const user = userEvent.setup()

    const {getByTestId, getByText, asFragment} = render(<Collapsable
      tasksList={tasksList}/>);

    const collapsibleDetails = getByTestId('collapsible-details');
    const collapsibleDetailsHeader = getByText('Test Tasks');

    await user.click(collapsibleDetailsHeader);
    expect(collapsibleDetails.getAttribute('open')).toBe('');

    await user.click(collapsibleDetailsHeader);
    expect(collapsibleDetails.getAttribute('open')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
});
