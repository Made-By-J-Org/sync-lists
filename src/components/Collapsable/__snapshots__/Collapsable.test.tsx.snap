// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collapsable > match snapshot 1`] = `
<DocumentFragment>
  <details
    data-testid="collapsible-details"
  >
    <summary>
      <span>
        <svg
          fill="none"
          height="15px"
          viewBox="0 0 24 24"
          width="15px"
        >
          <path
            d="M9 5H7C5.89543 5 5 5.89543 5 7V19C5 20.1046 5.89543 21 7 21H17C18.1046 21 19 20.1046 19 19V7C19 5.89543 18.1046 5 17 5H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M12 12H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M12 16H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M9 5C9 3.89543 9.89543 3 11 3H13C14.1046 3 15 3.89543 15 5V7H9V5Z"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <circle
            cx="9"
            cy="12"
            fill="#000000"
            r="1"
          />
          <circle
            cx="9"
            cy="16"
            fill="#000000"
            r="1"
          />
        </svg>
         
        <b>
          Test Tasks
        </b>
      </span>
      <span
        class="
            
            _positionText_6019c9
          "
      >
        <svg
          fill="#000"
          height="10px"
          version="1.1"
          viewBox="0 0 407 407"
          width="10px"
        >
          <polygon
            points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
          />
        </svg>
         Show
      </span>
    </summary>
    <div
      class="_content_6019c9"
    >
      <div
        class="_tasksGroupWrapper_6e408e"
      >
        <div
          class="_taskWrapper_416c64"
          data-testid="task-wrapper"
        >
          <span
            class="
          
          _checkbox_416c64
        "
            data-testid="task-wrapper-checkbox"
          />
          <span>
            Test Task
          </span>
        </div>
        <div
          class="_taskWrapper_416c64"
          data-testid="task-wrapper"
        >
          <span
            class="
          
          _checkbox_416c64
        "
            data-testid="task-wrapper-checkbox"
          />
          <span>
            Test Task2
          </span>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`Collapsable > toggles the state when details is clicked twice 1`] = `
<DocumentFragment>
  <details
    data-testid="collapsible-details"
  >
    <summary>
      <span>
        <svg
          fill="none"
          height="15px"
          viewBox="0 0 24 24"
          width="15px"
        >
          <path
            d="M9 5H7C5.89543 5 5 5.89543 5 7V19C5 20.1046 5.89543 21 7 21H17C18.1046 21 19 20.1046 19 19V7C19 5.89543 18.1046 5 17 5H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M12 12H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M12 16H15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M9 5C9 3.89543 9.89543 3 11 3H13C14.1046 3 15 3.89543 15 5V7H9V5Z"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <circle
            cx="9"
            cy="12"
            fill="#000000"
            r="1"
          />
          <circle
            cx="9"
            cy="16"
            fill="#000000"
            r="1"
          />
        </svg>
         
        <b>
          Test Tasks
        </b>
      </span>
      <span
        class="
            _rotateIcon_6019c9
            _positionText_6019c9
          "
      >
        <svg
          fill="#000"
          height="10px"
          version="1.1"
          viewBox="0 0 407 407"
          width="10px"
        >
          <polygon
            points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
          />
        </svg>
         Hide
      </span>
    </summary>
    <div
      class="_content_6019c9"
    >
      <div
        class="_tasksGroupWrapper_6e408e"
      >
        <div
          class="_taskWrapper_416c64"
          data-testid="task-wrapper"
        >
          <span
            class="
          
          _checkbox_416c64
        "
            data-testid="task-wrapper-checkbox"
          />
          <span>
            Test Task
          </span>
        </div>
        <div
          class="_taskWrapper_416c64"
          data-testid="task-wrapper"
        >
          <span
            class="
          
          _checkbox_416c64
        "
            data-testid="task-wrapper-checkbox"
          />
          <span>
            Test Task2
          </span>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;
